Adding capabilities document (#461)
microsoft/verona: docs/internal/capabilities.md

Regions are portions of memory dominated by an isolated (`iso`) pointer, with any number of mutable (`mut`) pointers in the same region, accessible only through the `iso` pointer.
Regions can be implemented in different ways, with different semantics.

For example, they can have different memory management strategies:
* Garbage collected: Internal objects are created, used and marked for destruction, with a collector freeing resources later.
* Reference counted: Objects have internal counts for each reference and when the count reaches zero, they're immediately deallocated.
* Arena model: Objects are allocated and used, but only freed on the destruction of the region itself.

They can also refer to how the code is executed:
* Native: Verona code calling Verona code, fully checked and safe.
* Sandbox: Verona code calling an external language module (ex. a C++ library), with process isolation.

Sandboxes themselves can also be of different types, for example:
* Child process: Forks a new process, shared memory, calls through RPC.
* WebASM: Isolates OS functionality, memory access, etc.

> * Remote execution: For use with different technologies (ex. GPU) or architectures (ex. connected embedded board).

Author: I can remove this for now.

None of those lists are complete, they're just examples of what can be done.

## Assignment

Capabilities change the semantics of assignments in the language.

**Isolated objects** can only be used through a single reference, so assigning an `iso` region to a new variable moves the reference and invalidates the previous one.

```ts
var foo : Foo & iso = Foo; // plus undecided syntax for "in new region type X"
foo.doSomething();
...

> var bar = foo; // bar now holds the *only* reference to foo

Author: Not sure what the example would be, as I'm not entirely clear the semantics of that issue. Happy to take suggestions.

> var bar = foo; // bar now holds the *only* reference to foo

Author: Do you mean that if a field is an `iso` and you assign to some other variable, the field will be left `null` and therefore not safe?

Like:
```
class Foo { field : Class & iso; create() { ... } };
...
let foo : Foo & mut = Foo; // now field has an isolated reference
let bar : Class & iso = foo.field; // ERROR: field cannot be nulled
```

bar.somethingElse(); // OK: bar is a reference to the same object as foo
foo.cleanup(); // ERROR: foo is not a valid reference any more
```

> **Mutable objects** can alias each other, so assigning a `mut` variable to another creates a (different) reference to the same variable.

Author: This is where it gets fuzzy for me, because the stack has a different lifetime than regions.

There are two issues here:
1. Compile time guarantees by the type system: if the object is never accessed after it's invalidated (for whatever reason), the code is "correct" and regardless of how we implement that at run-time, it's fine.
2. Run-time semantics: if the compiler said no one access the object after it's been made invalid, then stack lifetime is irrelevant, ie. all stack frames can have the "arena" model.

What cannot happen is functions trying to access objects in other frames that they don't own. Previous frames could be fine because their lifetime is strictly longer than the current's, but it'd be up for the compiler to prove that the access is correct via its own region's lifetime (which I'm not sure it's always possible).

Given that mutable references can only be modified by the one behaviour that holds their dominating `iso` reference, there are no race conditions in any write or reassignment.

```ts

> var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10

Author: Without a syntax for creating regions, it's really hard to write examples.

> var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10

Reviewer: We have syntax for creating regions: `new` and functions that wrap `new` such as `create`.  These always return an object in a new region.  We don't have syntax for specifying any constraints on the region type.

> var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10

Author: Right, I've been using `create` (assuming they'd wrap `new`) as object creation in a new region and `create ... in` for new object on an existing region. I don't think that's a reasonable syntax but I don't have a better one.

In our first MLIR dialect we had a distinction between `create region` and `create object`, which meant respectively: new object in a new region and new object in an existing region.

But none of that takes into account stack objects. If we want to create an object on the stack and `create` calls `new`, then do we need to create another method to do the same thing but not call `new`?

> var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10

Reviewer: I think objects created on the stack are in the abstract machine as objects created in a new region and then merged with the stack-frame region.  In the implementation, this does one of two things:

 - Allocates them with a heap allocator that cleans up on exit from the stack.
 - Promotes them to the stack.

We don't explicitly allocate objects on the stack, we allocate objects whose lifetimes are lexically scoped and let the compiler figure it out.

print(foo.value); // prints "10"
...
// Binding a new variable to the same reference creates a new reference to the same object
var bar = foo;
bar.value = 42;
print(foo.value); // prints "42"
...
// Creating a new object reassigns the variable bar but doesn't change the previous references (foo)
bar = Foo::create(31); // Returns a new Foo object with value = 31, binds it to bar
print(bar.value); // prints "31"
// But foo still points to the previous object
print(foo.value); // still prints "42"
```

**Read-only references** can be created from other capabilities, but cannot be assigned back to mutable references later.

This is to protect passing a read-only reference and the user change it back to a mutable one.
```ts
var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10
print(foo.value); // prints "10"
...
var bar : Foo & readonly = foo; // OK: a read-only reference to a mutable object
...
var baz : Foo & mut = bar; // ERROR: cannot convert readonly to mutable
```

> **Immutable objects** can also alias each other, but since none of those references can write through them, once created (frozen), an immutable region can only be read from.

Reviewer: ```suggestion
**Immutable capabilities enforce global immutability**.
The `freeze` operation consumes an `iso` capability.
The type system therefore guarantees that there can be no pointers to objects in the region (or any child regions) anywhere other than as pointers within the regions being frozen.
The `freeze` operation returns an `imm` capability and *viewpoint adaptation* guarantees that this cannot be used to materialise any kind of mutable capability.
Once a region has been frozen, no capabilities exist that permit writes to any objects that were in that region.
```

Author: Should we explain or link to a page explaining "viewpoint adaptation"? The first hits on search engines are all about computer vision and object detection, which can be super hard to filter.

Also, the Pony tutorial talks about viewpoint adaptation, but the Pony types are so much more complex that I'm not sure how we'd relate in the same way.

```ts
var foo : Foo & imm = freeze(Foo::create(10));
print(foo.value); // prints "10"
...
var bar = foo;
print(bar.value); // prints "10"
...
var ro : Foo & readonly = bar; // OK: imm is read-only anyway, so a noop
var imm : Foo & imm = ro; // OK? also a noop but there's a sub-type relationship that I'm not sure is there
```

### Destructive reads

Some reads invalidate the previous reference on assignment.
By design, `iso` references have a destructive read, since the program cannot have two `iso` references to the same region.

> But behaviour captures also destroy the previous reference to avoid race conditions.

Reviewer: ```suggestion
But behaviour captures also destroy the previous reference to avoid data races.
```

} // Schedules the behaviour and return -before- the execution finishes
...
// If this was allowed, there could be a race condition with _field0
foo.somethingElse(); // ERROR: foo was captured into a new reference and this one is now invalid
```

## Capability Conversions

Verona does not permit implicit type conversions and that also applies to capabilities.
Using a variable with one capability in place of another will not coerce the compiler to convert capabilities, unless there is a clear sub-type relationship or conversion function (ex. `freeze`).

### Sub-typing

Sub-typing is the type relationship where a sub-type can be safely used in place of its super type, for example, as function arguments or return values.
Sub-types can, therefore, be passed as arguments or directly assigned to variables that have been declared as their super-types.

Example:
```ts
interface Super {}
class Sub : Super {}
function(arg : Super) { ... }
...
a : Sub = Sub;
function(a); // OK, as Sub is a sub-type of Super
```

The notation is `Sub <: Super` that means `Sub` is a sub-type if `Super`.

### Allowed Conversions

These are the following capability sub-type relationships in Verona:

> * `iso <: mut`: Passing an `iso` in a (behaviour-local) place expecting a `mut` will _view_ the type as a mutable reference, because `iso` is strictly stringer than `mut`.

Reviewer: ```suggestion
* `iso <: mut`: An `iso` capability grants the rights to access a sentinel object and to create `mut` capabilities to any object in that region (i.e. that are reachable by following only `mut` pointers from the sentinel object) that are valid for as long as you hold the `iso` capability.
  As such, an `iso` capability trivially authorises constructing a `mut` capability to the object (this is the degenerate case: follow 0 `mut` pointers from the sentinel).
  As with any other `mut` capability, this may be stored only in places where the `iso` capability can be statically proven to be held: in fields of objects in the same region or in lexically-scoped variables dominated by a variable holding the `iso`.
  For example, if one stack frame holds an `iso` capability conferring exclusive ownership of a region then a function called from that stack frame may take a `mut` view of the sentinel object as an argument.
```

Author: Some of this is redundant with the description of `iso` and `sentinel objects` above. Perhaps we could elide some of the repetition here to just talk about the sub-type relationship?

> * `mut <: readonly`: Passing a `mut` in a place expecting a `readonly` will _view_ the type as read-only and not allow writes to it.

Reviewer: ```suggestion
* `mut <: readonly`: Passing a `mut` in a place expecting a `readonly` will _view_ the type as read-only and not allow writes to it.
  The `readonly` capability has the same lifetime restrictions as the `mut` capability from which it was derived.
```

> * `imm <: readonly`:  Similar reasoning, but here it's a no-op, because the object was globally immutable to begin with.

Reviewer: ```suggestion
* `imm <: readonly`:  An `imm` capability guarantees global immutability, `readonly` guarantees local immutability which is a strictly weaker property that is trivially satisfied by the stronger property.
```

I don't like the 'no-op' framing because that implies that the other conversions are non-trivial operations.

> When the conversion between capabilities require more than just a cast, we must call an explicit builtin to perform that functionality.

Reviewer: ```suggestion
Capability systems require that rights associated with a capability are never elevated except by interacting with a component that already owns these rights.
Any change to a capability involves an upgrade of rights or exchanging the existing rights for a mutually exclusive set then this must be an explicit operation.
```

* `iso -> imm` via `freeze`: Converts an `iso` region into an `imm` region by freezing it.

> The region will become globally immutable and cannot be unfrozen later, to avoid race conditions.

Reviewer: ```suggestion
The region will become globally immutable and cannot be unfrozen later, to guarantee data-race freedom.
```

# Capabilities

[Capabilities](https://en.wikipedia.org/wiki/Capability-based_security) are unforgeable tokens of authority that confer the rights to access something.
They can be delegated to other components of a system and must always respect the *principle of intentionality*: you must present a capability that confers the rights to perform an action when attempting to perform that action, it is not sufficient to simply own the capability.

In Verona, pointers are capabilities: the ability to name an object confers the right to access the object in some way.
Capabilities confer a range of different permissions, which are reflected in the type system.

Variables represent pointers to objects and they must have rights associated that describes what you can do to the objects via those pointers.

There are kinds of rights that can be granted to capabilities:
* Isolated (`iso`).

>   Defines the entry point to a region.
>   Whichever behaviour holds a reference to a region has the guarantee that no other behaviour has access to the same region.
>   To assign an `iso` reference to another variable is to transfer the ownership of that region.

Reviewer 2: ```suggestion
  Indicates that a capability is *the globally unique* pointer to its target.
  Every `iso` capability is the entry point to a region.
  Whichever running behaviour holds an `iso` reference to a region has the guarantee that no other running behaviour has any access to that region.
  To assign an `iso` reference to another variable *moves* it, removing it from the source and transfering the ownership of the target region.
```
